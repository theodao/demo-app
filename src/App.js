import React, {useState, useEffect} from "react";
import api from "./api";
import styled from "styled-components";
import ProgressBar from "./components/ProgressBar";
import Flex from "./components/styled/Flex";
import Button from "./components/Button";
import Container from "./components/Container";

const Headline = styled.h1``;

function App() {
  let [buttonValues, setButtonValues] = useState([]);
  let [barValues, setBarValues] = useState([]);
  let [loading, setLoading] = useState(false);
  let [limit, setLimit] = useState(0);
  let [currentIndex, setCurrentIndex] = useState(0);

  useEffect(() => {
    setLoading(true);
    api
      .getProgressBarData()
      .then(response => {
        const {buttons, bars, limit} = response;
        setButtonValues(buttons);
        setBarValues(bars);
        setLoading(false);
        setLimit(limit);
      })
      .catch(err => {
        console.log(err);
        setLoading(false);
      });
  }, []);

  const handleButtonClick = num => {
    setBarValues(
      barValues.map((item, index) => {
        if (index === currentIndex) {
          if (num + item > limit) {
            return limit;
          } else if (num + item < 0) {
            return 0;
          } else {
            return num + item;
          }
        } else {
          return item;
        }
      }),
    );
  };

  return (
    <Container loading={loading}>
      <Flex justifyContent="center" alignItems="center" style={{height: "100vh"}}>
        <div>
          <Headline>Progress Bars Demo</Headline>
          {/* Should not set key as index here */}
          <Flex>
            {barValues.map((barValue, index) => {
              return <ProgressBar percentage={barValue} color="#6497b1" limit={limit} key={index} />;
            })}
          </Flex>
          <Flex flexDirection="row" justifyContent="space-between">
            {buttonValues.map((buttonValue, index) => {
              return (
                <Button
                  key={index}
                  onClick={() => {
                    handleButtonClick(buttonValue);
                  }}>
                  {buttonValue}
                </Button>
              );
            })}
          </Flex>
          <div style={{margin: "10px"}}></div>
          <select value={currentIndex} onChange={e => setCurrentIndex(Number(e.target.value))}>
            {barValues.map((barValue, index) => {
              return <option key={index} value={index}>{`Progress bar #${index + 1}`}</option>;
            })}
          </select>
        </div>
      </Flex>
    </Container>
  );
}

export default App;
