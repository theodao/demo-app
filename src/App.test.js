import React, {useState as useStateMock} from "react";
import renderer from "react-test-renderer";
import App from "./App";
import {shallow} from "enzyme";

jest.mock("./api", () => {
  return {
    getProgressBarData: () => {
      return new Promise((resolve) => {
        resolve(["data"]);
      });
    },
  };
});

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
  useRef: jest.fn(),
}));

const setState = jest.fn();

describe("Test case for App.js", () => {
  let wrapper = null;

  beforeEach(() => {
    useStateMock.mockImplementation(init => [init, setState]);
    wrapper = shallow(<App />);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders App", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("call api", () => {
    const tree = renderer.create(<App />);
    tree.update();
    expect(setState).toHaveBeenCalled();
  });

  it("option does trigger on change", () => {
    wrapper
      .find("select")
      .props()
      .onChange({
        target: {
          value: "1",
        },
      });
    expect(setState).toHaveBeenCalled();
  });
});
