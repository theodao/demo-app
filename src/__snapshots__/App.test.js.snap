// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for App.js renders App 1`] = `
<div>
  <div
    className="sc-AxhUy kcStfL"
    style={
      Object {
        "height": "100vh",
      }
    }
  >
    <div>
      <h1
        className="sc-AxmLO cAKCgh"
      >
        Progress Bars Demo
      </h1>
      <div
        className="sc-AxhUy gScyUb"
      />
      <div
        className="sc-AxhUy ivmIcr"
      />
      <div
        style={
          Object {
            "margin": "10px",
          }
        }
      />
      <select
        onChange={[Function]}
        value={0}
      />
    </div>
  </div>
</div>
`;
